### PR TITLE
TF 1.15 compatibility

### PR DESCRIPTION
**Pull request by pzelasko**
Work in progress.

### CODE REVIEW

File: deps/install_tensorflow_1_15.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+pip install tensorflow==1.15.0
+
+# Find out where do the tensorflow libs and sources reside
+TF_DIR="$(python -c 'import tensorflow as tf; print(tf.sysconfig.get_lib())')"
+
+# Make TF lib detectable for the ldd linker
+ln -s "$TF_DIR/libtensorflow_framework.so.1" "$TF_DIR/libtensorflow_framework.so"
+
+# Work around TF issue: https://github.com/tensorflow/tensorflow/issues/22766
+sed -i 's/return ABSL_ASSERT(len <= kMaxSize), len;/return len;/' "$TF_DIR/include/absl/strings/string_view.h"
```
Comment (albertz) [inline deps/install_tensorflow_1_15.sh]: As far as I remember, this was only needed for GPU? In that case, we should not do that  for Travis.

File: deps/install_tensorflow_1_15.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+set -eou pipefail
+
+pip install tensorflow==1.15.0
+
+# Find out where do the tensorflow libs and sources reside
+TF_DIR="$(python -c 'import tensorflow as tf; print(tf.sysconfig.get_lib())')"
+
+# Make TF lib detectable for the ldd linker
+ln -s "$TF_DIR/libtensorflow_framework.so.1" "$TF_DIR/libtensorflow_framework.so"
```
Comment (albertz) [inline deps/install_tensorflow_1_15.sh]: Why is this needed? So everyone else also needs to do this? Shouldn't it be simple to just fix the compile/ld flags at compilation to directly find the right so-file? I think this would be the better solution.

File: .travis.yml [hunk 3]
```diff
@@ -26,7 +26,7 @@ install:
   - pip install -qq --upgrade pip setuptools wheel
   - pip install --upgrade numpy scipy | cat  # TF usually needs the latest Numpy
   - pip install -r requirements.txt | cat
-  - pip install $TF_PACKAGE | cat
+  - bash deps/install_tensorflow_1_15.sh
```
Comment (albertz) [inline .travis.yml]: It should install `$TF_PACKAGE` here, **not** TF 1.15.

Comment (albertz) [reply .travis.yml]: Also, because of that, the install Bash script is not so nice.
In any case, I would try to avoid using the Bash script here if possible. Maybe I will just enabled `-DNDEBUG` globally now, so that the patching should not be needed anymore. See also the discussion in #87.

File: NativeOp.cpp [hunk 4]
```diff
@@ -752,7 +752,11 @@ void* _malloc(size_t num_bytes) {
     //auto* stream = context->op_device_context()->stream();
     Allocator* allocator =
         context->device()->GetAllocator(AllocatorAttributes());
-    void* ptr = (void*)allocator->Allocate<uint8_t>(num_bytes);
+#if (TF_MAJOR_VERSION == 1 && TF_MINOR_VERSION >= 14)
```
Comment (albertz) [inline NativeOp.cpp]: Can you add maybe a comment? `Allocate` was removed from the API since TF 1.14?

File: TFNativeOp.py [hunk 5]
```diff
@@ -243,7 +243,12 @@ def make_dim_str(c):
     for in_idx, v in enumerate(in_info):
       out_idx = v.get("want_inplace", -1)
       if out_idx >= 0:
-        code_forward_io += "context->forward_ref_input_to_ref_output(%i, %i);\n" % (in_idx, out_idx)
+        code_forward_io += """
+if (IsRefType(context->input_dtype({in_idx})))
+  context->forward_ref_input_to_ref_output({in_idx}, {out_idx});
+else
+  context->set_output({out_idx}, context->input({in_idx}));
+\n""".format(in_idx=in_idx, out_idx=out_idx)
```
Comment (albertz) [inline TFNativeOp.py]: Why the `\n`? Also, indent it to the other Python code, which looks a bit nicer in the Python code (I don't care much how the resulting C++ code is indented...).

Comment (albertz) [reply TFNativeOp.py]: Btw, also add a comment here. Did you investigate further what this actually means, and why it happens now, and not earlier?
@curufinwe You said in your compilation of TF 1.15, someone observed some strange behavior in training. Maybe this is related?

File: deps/install_tensorflow_1_15.sh (hunk 1 above)
Comment (albertz) [reply deps/install_tensorflow_1_15.sh]: (#87 says this is for nvcc only...)

File: requirements.txt [hunk 6]
```diff
@@ -1,3 +1,10 @@
...
 h5py
+nose
+numpy
+scipy
+setuptools
+six
+typing
 typing
+wheel
+
```
Comment (albertz) [inline requirements.txt]: Don't include this in this pull request.
Also, in general, this should not be needed. Esp it should not contain stuff which are only needed for the tests.
Numpy should also be covered already when you install TF (TF is not in there because some people would use `tensorflow-gpu`, while others use just `tensorflow`).
Scipy is not necessarily needed.
We don't use Six (I think...). Anyway, I guess also already covered implicitly by TF.
Setuptools, Wheels also not needed.
Typing is twice there now.

File: .travis.yml (hunk 3 above)
Comment (albertz) [reply .travis.yml]: Btw, also check [Travis](https://travis-ci.org/rwth-i6/returnn/builds/621529097): It seems it works fine without this.

File: deps/install_tensorflow_1_15.sh [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
Comment (albertz) [inline deps/install_tensorflow_1_15.sh]: See the comment above. I think we don't need the file.

File: requirements.txt (hunk 6 above)
Comment (pzelasko) [reply requirements.txt]: Sure, I can revert these. I had some test failures when scipy was not installed though, so perhaps it should be included there?

Comment (albertz) [reply requirements.txt]: Where exactly? But anyway, let's move this out of this pull request. Make a separate issue maybe first, including the exception you get/got.

File: TFNativeOp.py (hunk 5 above)
Comment (pzelasko) [reply TFNativeOp.py]: I'll fix the formatting. 

The `forward_ref...` bit expects to have tensor references both on input and output, and sets `output = input`.  It seems that this assumption was not satisfied in some tests, i.e. the input was actually a tensor value, not a reference, which triggered assertion.

Unfortunately, I don't know why it started happening only with 1.14/1.15... Didn't find any relevant changes in the TF change logs.

File: NativeOp.cpp (hunk 4 above)
Comment (pzelasko) [reply NativeOp.cpp]: That's right. I'll add a comment.

File: .travis.yml (hunk 3 above)
Comment (pzelasko) [reply .travis.yml]: OK will do.

File: tests/test_TFEngine.py [hunk 8]
```diff
@@ -391,15 +391,6 @@ def test_engine_train_grad_noise_sparse():
...
 
-  def wrapped_slice_to_tensor(value, dtype=None, name=None, as_ref=False):
-    print("wrapped_slice_to_tensor:", value)
-    #assert "flatten_with_seq_len_mask" in value.name
-    from tensorflow.python.ops import gradients_impl
-    return gradients_impl._IndexedSlicesToTensor(value, dtype=dtype, name=name, as_ref=as_ref)
-
-  from tensorflow.python.framework import ops
-  ops.register_tensor_conversion_function(tf.IndexedSlices, wrapped_slice_to_tensor, priority=1)
-
```
Comment (albertz) [inline tests/test_TFEngine.py]: Why only removing this? How does it fix sth?
Also, there was no problem in Travis. Why is this needed?

File: deps/install_tensorflow_1_15.sh (hunk 2 above)
Comment (pzelasko) [reply deps/install_tensorflow_1_15.sh]: Let me check again then - I had no success with this approach earlier.

File: deps/install_tensorflow_1_15.sh (hunk 1 above)
Comment (pzelasko) [reply deps/install_tensorflow_1_15.sh]: OK I'll try to remove the Bash script altogether.

File: tests/test_TFEngine.py (hunk 8 above)
Comment (pzelasko) [reply tests/test_TFEngine.py]: They removed the `_IndexedSlicesToTensor` function from `gradients_impl.py`, I think they replaced it's functionality with something else. The test didn't pass when I checked locally, after the change it works fine.

File: deps/install_tensorflow_1_15.sh (hunk 7 above)
Comment (pzelasko) [reply deps/install_tensorflow_1_15.sh]: Ok

File: TFNativeOp.py (hunk 5 above)
Comment (albertz) [reply TFNativeOp.py]: Yea, don't explain here, but in the code with a comment.
Also, what does it mean to be a reference or a value?

**Comment by albertz** [comment]
I see that the `DebugString()` API changed. It became `const` now. So we need some
```
#if (TF_MAJOR_VERSION == 1 && TF_MINOR_VERSION <= 13)
const
#endif 
```
there. (I think this is relevant for OpenFst and KenLM only.)

**Comment by albertz** [comment]
Btw, also maybe cleanup the commits already (rebase, squash together a bit).

File: tests/test_TFEngine.py (hunk 8 above)
Comment (albertz) [reply tests/test_TFEngine.py]: I see, ok.

**Comment by pzelasko** [comment]
Closing this PR; there is a new one with commits squashed: #251

**Comment by albertz** [comment]
> Closing this PR; there is a new one with commits squashed: #251

It's a bad idea to do that. You should rather just rebase and force-push/update this pull request. Otherwise we lose all the discussion here.

**Comment by pzelasko** [comment]
You're right!

File: TFUtil.py [hunk 9]
```diff
@@ -4787,7 +4787,7 @@ def __init__(self, use_cuda_if_available=True, cuda_auto_min_compute_capability=
     ld_flags = list(ld_flags)
     if have_min_tf_version((1, 4)):
       # https://github.com/tensorflow/tensorflow/issues/13607
-      ld_flags += ["-L%s" % tf.sysconfig.get_lib(), "-ltensorflow_framework"]
+      ld_flags += tf.sysconfig.get_link_flags()
```
Comment (albertz) [inline TFUtil.py]: Be careful here. I think `get_link_flags()` was introduced only later (later than TF 1.4). So maybe we need another check `have_min_tf_version((1, 14))` (or whatever version) for that.

File: TFUtil.py [hunk 10]
```diff
@@ -8071,6 +8078,19 @@ def tensor_array_is_clear_after_read(ta):
 
 # noinspection PyProtectedMember
 def tensor_array_element_shape(ta):
+  """
+  :param tf.TensorArray ta:
+  :rtype: tf.TensorShape
+  """
+  tf_version = float(tf.version.VERSION.rsplit('.', 1)[0])
+  if tf_version >= 1.14:
```
Comment (albertz) [inline TFUtil.py]: Use `have_min_tf_version((1, 14))` instead.

File: TFUtil.py [hunk 11]
```diff
@@ -8085,6 +8105,15 @@ def tensor_array_element_shape(ta):
   return tf.TensorShape(None)
 
 
+def tensor_array_element_shape_py3_tf_1_14_and_later(ta):
+  """
+  :param tf.TensorArray ta:
+  :rtype: tf.TensorShape
+  """
+  assert isinstance(ta.element_shape, tf.TensorShape)
+  return ta.element_shape
```
Comment (albertz) [inline TFUtil.py]: Do not create an extra function for this. Just make it inline in `tensor_array_element_shape`. The same also for `tensor_array_element_shape_tf_up_to_1_13`.

File: requirements.txt [hunk 12]
```diff
@@ -1,3 +1,3 @@
-numpy
 h5py
+numpy
```
Comment (albertz) [inline requirements.txt]: Put it back to top, there should be no change here.

File: tests/test_TFEngine.py [hunk 13]
```diff
@@ -391,14 +391,16 @@ def test_engine_train_grad_noise_sparse():
...
-  def wrapped_slice_to_tensor(value, dtype=None, name=None, as_ref=False):
-    print("wrapped_slice_to_tensor:", value)
-    #assert "flatten_with_seq_len_mask" in value.name
-    from tensorflow.python.ops import gradients_impl
-    return gradients_impl._IndexedSlicesToTensor(value, dtype=dtype, name=name, as_ref=as_ref)
-
-  from tensorflow.python.framework import ops
-  ops.register_tensor_conversion_function(tf.IndexedSlices, wrapped_slice_to_tensor, priority=1)
+  tf_version = float(tf.version.VERSION.rsplit('.', maxsplit=1)[0])  # e.g. convert "1.15.0" -> 1.15
+  if tf_version < 1.14:
```
Comment (albertz) [inline tests/test_TFEngine.py]: Use `have_min_tf_version` instead.

File: TFNativeOp.py [hunk 14]
```diff
@@ -243,7 +243,12 @@ def make_dim_str(c):
     for in_idx, v in enumerate(in_info):
       out_idx = v.get("want_inplace", -1)
       if out_idx >= 0:
-        code_forward_io += "context->forward_ref_input_to_ref_output(%i, %i);\n" % (in_idx, out_idx)
+        code_forward_io += """
+          // Note: forward_ref_input_to_ref_output expects the input tensor to be a reference, and not a value;
+          //   otherwise it will trigger an assertion.
+          if (IsRefType(context->input_dtype({in_idx})))
+            context->forward_ref_input_to_ref_output({in_idx}, {out_idx});
```
Comment (albertz) [inline TFNativeOp.py]: Now you removed the `else` part. Is that right?

File: tests/test_TFEngine.py [hunk 15]
```diff
@@ -391,14 +391,16 @@ def test_engine_train_grad_noise_sparse():
...
+    def wrapped_slice_to_tensor(value, dtype=None, name=None, as_ref=False):
+      print("wrapped_slice_to_tensor:", value)
+      #assert "flatten_with_seq_len_mask" in value.name
+      from tensorflow.python.ops import gradients_impl
+      return gradients_impl._IndexedSlicesToTensor(value, dtype=dtype, name=name, as_ref=as_ref)
+
+    from tensorflow.python.framework import ops
+    ops.register_tensor_conversion_function(tf.IndexedSlices, wrapped_slice_to_tensor, priority=1)
```
Comment (albertz) [inline tests/test_TFEngine.py]: I think you can just remove the code. There anyway was just some debug `print` statement.

File: TFUtil.py [hunk 16]
```diff
@@ -4850,7 +4850,7 @@ def __init__(self, include_paths=(), ld_flags=(), **kwargs):
     ld_flags = list(ld_flags)
     if have_min_tf_version((1, 4)):
       # https://github.com/tensorflow/tensorflow/issues/13607
-      ld_flags += ["-L%s" % tf.sysconfig.get_lib(), "-ltensorflow_framework"]
+      ld_flags += tf.sysconfig.get_link_flags()
```
Comment (albertz) [inline TFUtil.py]: Be careful here. I think `get_link_flags()` was introduced only later (later than TF 1.4). So maybe we need another check `have_min_tf_version((1, 14))` (or whatever version) for that.

File: TFNativeOp.py (hunk 14 above)
Comment (pzelasko) [reply TFNativeOp.py]: Yep - otherwise, TF quits with error and complains about more than one allocation on the output tensor. With the current state, all the tests seem to pass.

**Comment by albertz** [comment]
Now you have it all in one single commit, which is not so optimal. It should better be split into multiple commits. (I just requested to cleanup the history, i.e. to squash it a bit, but not into a single commit).

File: TFNativeOp.py (hunk 14 above)
Comment (albertz) [reply TFNativeOp.py]: Ah ok I see. Still not sure if that `forward_ref_input_to_ref_output` makes sense actually... But anyway, maybe we can leave it like that for now.

**Comment by pzelasko** [comment]
Got it. Do you know if this can be reverted somehow? How many commits would you suggest?

**Comment by albertz** [comment]
Undoing that is a bit tricky. See e.g. [here](https://stackoverflow.com/questions/6217156/break-a-previous-commit-into-multiple-commits).

**Comment by albertz** [comment]
Btw, you also tested that on GPU? It works fine (when you work around the nvcc absl problem manually somehow)?

File: .travis.yml [hunk 17]
```diff
@@ -41,8 +41,7 @@ install:
 # TF tests first, because more relevant (for me), then ordered by alphabet
 env:
   global:
-    # TODO: TF 1.14 needs multiple fixes...
-    - TF_PACKAGE=tensorflow==1.13.1
```
Comment (albertz) [inline .travis.yml]: I would move this change to an own commit, and it should be the last commit in this PR.

Comment (pzelasko) [reply .travis.yml]: No problem!

**Comment by pzelasko** [comment]
Haven't tested on GPU yet. Let me see how it works.

**Comment by albertz** [comment]
Tests looking good, and I don't think there is any reason to not merge this in as it is now.
Thank you very much for the work!

If there are any problems which you (also @curufinwe @Spotlight0xff @pavelgolik @critias and others) observe later for TF 1.14 or TF 1.15, please report a separate issue.
